src/server/lib/ophan.ts

```diff
@@ -2,17 +2,123 @@ import { OphanEvent, OphanInteraction } from '@/shared/model/ophan';
 import { fetch } from '@/server/lib/fetch';
 import { logger } from '@/server/lib/serverSideLogger';
 import timeoutSignal from 'timeout-signal';
-import { stringify } from 'query-string';
+import { stringify, parse } from 'query-string';
+import { z } from 'zod';
+import {
+  OphanAction,
+  OphanComponent,
+  OphanComponentEvent,
+  OphanComponentType,
+} from '@guardian/libs';
+import serialize from 'serialize-javascript';
 
 const ophanUrl = 'https://ophan.theguardian.com/img/2';
 
+// Component event params is the decoded version of the componentEventParams query parameter
+// defined in the TrackingQueryParams interface from src/shared/model/QueryParams.ts
+const componentEventParamsSchema = z
+  .object({
+    componentType: z.string(),
+    componentId: z.string(),
+    abTestName: z.string().optional(),
+    abTestVariant: z.string().optional(),
+    viewId: z.string(),
+    browserId: z.string().optional(),
+    visitId: z.string().optional(),
+  })
+  .strict();
+
+type ComponentEventParams = z.infer<typeof componentEventParamsSchema>;
+
+/**
+ * OphanConfig is the configuration for an Ophan event,
+ * for example the browser id (bwid), the consent UUID (consentUUID)
+ * and the page view id (viewId)
+ */
 export interface OphanConfig {
   bwid?: string;
   viewId?: string;
   consentUUID?: string;
 }
 
-export const record = (event: OphanEvent, config: OphanConfig = {}) => {
+/**
+ * Convert the componentEventParams query string to the
+ * ComponentEventParams type using zod, return undefined if
+ * unable to parse the query string
+ */
+const parseComponentEventParams = (
+  componentEventParams: string,
+): ComponentEventParams | undefined => {
+  const parsedQuery = parse(componentEventParams.replace('undefined', ''));
+
+  const componentEventParamsParsed =
+    componentEventParamsSchema.safeParse(parsedQuery);
```

Comment:
I wonder if [`safeParseAsync`](https://github.com/colinhacks/zod#safeparseasync) might be a better method to use here so we don't block the main thread, e.g if someone malicious decides to hit us with a lot of data?

Reply:
That's a very good point! Updated to use `safeParseAsync`.